### PR TITLE
[Optimize] Move dead partial apply elimination to mandatory combine

### PR DESCRIPTION
**Pull request by zoecarver**
This patch moves the removal of dead partial applies to mandatory combine (from mandatory inlining). SourceKit doesn't run mandatory combine but, it does run mandatory inlining, so the more we can move out of mandatory inlining, the better. 

refs #28484 #28394 #28396
cc @gottesmm

### CODE REVIEW

**Comment by zoecarver** [comment]
@gottesmm friendly ping. Does the general idea of this patch look good to you?

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp [hunk 1]
```diff
@@ -270,6 +271,23 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
...
+  // If there are more than one uses, we don't want to remove this.
+  if (!user)
+    return nullptr;
+  if (!isa<DeallocStackInst>(user->getUser()))
+    return nullptr;
+
+  // Remove the stack dealloc, then the partial apply, then the function ref.
+  instModCallbacks.deleteInst(user->getUser());
+  instModCallbacks.deleteInst(i);
+  instModCallbacks.deleteInst(cast<FunctionRefInst>(i->getCallee()));
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/MandatoryCombine.cpp]: It's unlikely but possible that the function_ref has other uses, so I'm not sure it's quite safe to remove it.

**Review by slavapestov — COMMENTED** [review]
Are there any tests that exercise the old code? If not, do you mind adding one?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by zoecarver** [comment]
@slavapestov this should mostly be a NFC. So, no tests should need to change. Do you mean I should add a test that only runs mandatory inlining (to make sure the closure isn't optimized)?

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp (hunk 1 above)
Comment (zoecarver) [reply lib/SILOptimizer/Mandatory/MandatoryCombine.cpp]: Good point. I'll fix this.

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp [hunk 2]
```diff
@@ -270,6 +271,25 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
...
+    return nullptr;
+  if (!isa<DeallocStackInst>(user->getUser()))
+    return nullptr;
+
+  // Remove the stack dealloc, then the partial apply, then the function ref.
+  instModCallbacks.deleteInst(user->getUser());
+  instModCallbacks.deleteInst(i);
+  // If the only use of the function_ref is us, then remove it.
+  if (i->getCallee()->getSingleUse()->getUser() == i)
+    instModCallbacks.deleteInst(cast<FunctionRefInst>(i->getCallee()));
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/MandatoryCombine.cpp]: I don't think you can guarantee that the callee is always going to be a function_ref. Can you add an isa check to make sure?

**Comment by slavapestov** [comment]
It's up to you if you want to add a SIL test that only runs mandatory combining or not.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/SILOptimizer/mandatory_inlining_and_combine.sil [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+// INLINE-CHECK: partial_apply
+bb0(%0 : $Builtin.Int8):
+  %2 = function_ref @outer_transparent : $@convention(thin) (@owned @callee_owned () -> Builtin.Int8) -> Builtin.Int8
+  %3 = function_ref @identity_closure : $@convention(thin) (Builtin.Int8) -> Builtin.Int8
+  %4 = partial_apply %3(%0) : $@convention(thin) (Builtin.Int8) -> Builtin.Int8
+  %5 = apply %2(%4) : $@convention(thin) (@owned @callee_owned () -> Builtin.Int8) -> Builtin.Int8
+  return %5 : $Builtin.Int8
+}
+
+sil shared [transparent] @identity_closure : $@convention(thin) (Builtin.Int8) -> Builtin.Int8 {
```
Comment (zoecarver) [inline test/SILOptimizer/mandatory_inlining_and_combine.sil]: I'm ignoring the fact that this function no longer gets removed because it seems inline with the update in #28484

File: test/SILOptimizer/mandatory_inlining.sil [hunk 4]
```diff
@@ -1088,7 +1012,7 @@ bb0(%0 : $C):
 // CHECK:   [[F:%.*]] = function_ref @use_c
 // CHECK:   apply [[F]]([[ARG]])
 // CHECK:   strong_release [[ARG]]
-// CHECK:   strong_release [[ARG]]
+// TODO: why is this broken?
```
Comment (zoecarver) [inline test/SILOptimizer/mandatory_inlining.sil]: I need to figure out why there are too few `strong_release`s here.

File: test/SILOptimizer/mandatory_inlining_and_combine.sil [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+// CHECK: [[CALL_BB]]:
+// CHECK-NEXT:  function_ref use_c
+// CHECK-NEXT:  [[FUNC:%.*]] = function_ref @use_c : $@convention(thin)
+// CHECK-NEXT:  apply [[FUNC]]
+// CHECK-NEXT:  tuple
+// CHECK-NEXT:  strong_release [[ARG]]
+// CHECK-NEXT:  br [[MERGE_BB]]
+//
+// CHECK: [[MERGE_BB]]:
+// TODO: strong_release [[ARG]]
```
Comment (zoecarver) [inline test/SILOptimizer/mandatory_inlining_and_combine.sil]: I was able to fix the other `strong_release` todos but I'm still looking into the cause of this one.

**Comment by zoecarver** [comment]
@slavapestov I've added `mandatory_inlining_and_combine.sil` to test that mandatory inlining and mandatory combine each do the "right" thing. 

I had to update this patch quite a bit to cover all the closure cleanup done in mandatory inlining, now it should cover all the cases that it used to. 

`mandatory_inlining_ownership2.sil` had a lot of tests that now required both a mandatory inlining and mandatory combine pass. Instead of moving them to `mandatory_inlining_and_combine.sil` I added a `RUN` comment that runs both passes. If you'd prefer, I'm happy to move it into `mandatory_inlining_and_combine.sil` (no sure which is best).

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 6]
```diff
@@ -1449,7 +1449,8 @@ static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
 
   ValueLifetimeAnalysis::Frontier partialApplyFrontier;
   if (!vla.computeFrontier(partialApplyFrontier,
-                           ValueLifetimeAnalysis::DontModifyCFG)) {
+                           ValueLifetimeAnalysis::DontModifyCFG) ||
+      partialApplyFrontier.empty()) {
```
Comment (zoecarver) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: If `partialApplyFrontier` is empty (and we don't return), we will create an `alloc_stack` and `copy_addr` but no `dealloc_stack`.

**Comment by zoecarver** [comment]
@slavapestov and @gottesmm friendly ping. I've made sure that all the tests will pass after this patch is applied. Some of the tests appear to have been deleted but that either reflects an increase in deleted `partial_apply`s or they have been moved elsewhere (such as from `mandatory_inlining.sil` to `mandatory_inlining_and_combine.sil`). 

I realize this is a large patch so it may take a while to review. Let me know if you have any questions.

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
@swift-ci please test.

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 7]
```diff
@@ -1943,3 +1944,102 @@ AbstractFunctionDecl *swift::getBaseMethod(AbstractFunctionDecl *FD) {
   }
   return FD;
 }
+
+SILValue swift::stripCopiesAndBorrows(SILValue v) {
+  while (isa<CopyValueInst>(v) || isa<BeginBorrowInst>(v)) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: I think we already have a routine like this. I would look around.

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp [hunk 8]
```diff
@@ -201,11 +203,6 @@ bool MandatoryCombiner::doOneIteration(SILFunction &function,
       );
     }
 
-    for (SILInstruction *instruction : instructionsPendingDeletion) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/MandatoryCombine.cpp]: Why is this removed?

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp [hunk 9]
```diff
@@ -284,6 +281,86 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
...
+    if (tryDeleteDeadClosure(tttf, instModCallbacks)) {
+      val = tttf->getCallee();
+      if (auto *ref = getRemovableRef(tttf)) {
+        instModCallbacks.deleteInst(ref);
+        return nullptr;
+      }
+    }
+  }
+
+  if (auto *convFunc = dyn_cast<ConvertFunctionInst>(val)) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/MandatoryCombine.cpp]: This is not needed. Should happen automatically when we go back around.

**Review by gottesmm — COMMENTED** [review]
Some quick comments. I am going to want to relook at this after you make improvements/what we talked about over chat.

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp [hunk 10]
```diff
@@ -284,6 +281,86 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
...
+  if (auto *pa = dyn_cast<PartialApplyInst>(src)) {
+    if (tryDeleteDeadClosure(pa, instModCallbacks)) {
+      val = pa->getCallee();
+      if (auto *ref = getRemovableRef(pa)) {
+        instModCallbacks.deleteInst(ref);
+        return nullptr;
+      }
+    }
+  } else if (auto *tttf = dyn_cast<ThinToThickFunctionInst>(src)) {
+    if (tryDeleteDeadClosure(tttf, instModCallbacks)) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/MandatoryCombine.cpp]: Most of the optimizations here should not be on LoadInst. You aren't removing load inst anywhere! Instead, all of these should be on the actual instructions you are removing (i.e. ThinToThick, Partial, Etc).

**Comment by zoecarver** [comment]
I'll make all these changes as their own (smaller, more focused) PRs.
